modules/consensus/accept.go

```diff
@@ -207,80 +217,153 @@ func (cs *ConsensusSet) addBlockToTree(b types.Block, parent *processedBlock) (c
-			// ban peers that send lots of future blocks and stop spawning goroutines
-			// after we are already waiting on a large number of future blocks.
-			//
-			// TODO: an attacker could broadcast a future block many times and we would
-			// spawn a goroutine for each broadcast. To fix this we should create a
-			// cache of future blocks, like we already do for DoS blocks, and only spawn
-			// a goroutine if we haven't already spawned one for that block. To limit
-			// the size of the cache of future blocks, make it a constant size (say 50)
-			// over which we would evict the block furthest in the future before adding
-			// a new block to the cache.
+	// Verify the headers for every block, throw out known blocks, and the
+	// invalid blocks (which includes the children of invalid blocks).
+	chainExtended := false
+	changes := make([]changeEntry, 0, len(blocks))
+	validBlocks := make([]types.Block, 0, len(blocks))
+	parents := make([]*processedBlock, 0, len(blocks))
+	setErr := cs.db.Update(func(tx *bolt.Tx) error {
+		for i := 0; i < len(blocks); i++ {
+			// Start by checking the header of the block.
+			parent, err := cs.validateHeaderAndBlock(boltTxWrapper{tx}, blocks[i], blockIDs[i])
+			if err == modules.ErrBlockKnown {
+				// Skip over known blocks.
+				continue
+			}
 			if err == errFutureTimestamp {
-				go func() {
-					time.Sleep(time.Duration(b.Timestamp-(types.CurrentTimestamp()+types.FutureThreshold)) * time.Second)
-					err := cs.managedAcceptBlock(b)
-					if err != nil {
-						cs.log.Debugln("WARN: failed to accept a future block:", err)
-					}
-					cs.managedBroadcastBlock(b)
-				}()
+				// Queue the block to be tried again if it is a future block.
+				go cs.threadedSleepOnFutureBlock(blocks[i])
+			}
+			if err != nil {
+				return err
+			}
+
+			// Try adding the block to consnesus.
+			changeEntry, err := cs.addBlockToTree(tx, blocks[i], parent)
+			if err == nil {
+				chainExtended = true
+			}
+			if err == modules.ErrNonExtendingBlock {
+				err = nil
 			}
-			return err
+			if err != nil {
+				return err
+			}
+			// Sanity check - If reverted blocks is zero, applied blocks should also
+			// be zero.
+			if build.DEBUG && len(changeEntry.AppliedBlocks) == 0 && len(changeEntry.RevertedBlocks) != 0 {
+				panic("after adding a change entry, there are no applied blocks but there are reverted blocks")
+			}
+			// Append to the set of changes, and append the valid block.
+			changes = append(changes, changeEntry)
+			validBlocks = append(validBlocks, blocks[i])
+			parents = append(parents, parent)
 		}
 		return nil
 	})
-	if err != nil {
-		cs.mu.Unlock()
-		return err
+	if setErr != nil {
+		// Check if any blocks were valid.
+		if len(validBlocks) < 1 {
+			// Nothing more to do, the first block was invalid.
+			return false, setErr
+		}
+
+		// At least some of the blocks were valid. Add the valid blocks before
+		// returning, since we've already done the downloading and header
+		// validation.
```

Comment:
is there a DoS attack wherein the attacker sends the maximum number of blocks, with the last being invalid, forcing consensus to redo a bunch of work?